### PR TITLE
Mobilecoind performance improvements

### PR DESCRIPTION
**Pull request by eranrund**
Soundtrack of this PR: [Chino Moreno - Brief Exchange](https://www.youtube.com/watch?v=sBMGiHosRYk)

### Motivation

Slow ledger syncing is a suboptimal user-experience, and we'd like it to be faster.

### In this PR
* Increasing number of worker threads when fetching S3 blocks, as well as increasing the amount of block headers fetched from consensus. This results in bringing ledger sync time from several minutes to 40-60 seconds on my machines.
* Switch a mutex around `NetworkState` to an `RwLock` - this allows mobilecoind's `GetNetworkStatus` RPC call to not block when a ledger sync is taking place.

### Future Work
* Dig further into why attempts at changing the S3 fetching to use async code result in worse performance than creating 500 real threads.
* Consider post-processing blocks into larger chunks in order to reduce number of S3 requests needed.

### CODE REVIEW

**Review by tsegaran — APPROVED** [review]
Awesome! Thanks for digging into this.

File: ledger/sync/src/ledger_sync/ledger_sync_service.rs [hunk 1]
```diff
@@ -33,6 +33,9 @@ use std::{
 const DEFAULT_GET_BLOCKS_TIMEOUT: Duration = Duration::from_secs(30);
 const DEFAULT_GET_TRANSACTIONS_TIMEOUT: Duration = Duration::from_secs(30);
 
+/// Maximal amount of concurrent get_block_contents calls to allow.
+const MAX_CONCURRENT_GET_BLOCK_CONTENTS_CALLS: usize = 500;
```
Comment (sugargoat) [inline ledger/sync/src/ledger_sync/ledger_sync_service.rs:37]: Should we ultimately expose this to make it configurable?

File: ledger/sync/src/ledger_sync/ledger_sync_service_thread.rs [hunk 2]
```diff
@@ -10,14 +10,14 @@ use mc_ledger_db::Ledger;
...
-        Arc, Mutex,
+        Arc, RwLock,
     },
     thread,
     time::Duration,
 };
 
 /// Maximal number of blocks to attempt to sync at each loop iteration.
-const MAX_BLOCKS_PER_SYNC_ITERATION: u32 = 100;
+const MAX_BLOCKS_PER_SYNC_ITERATION: u32 = 10000;
```
Comment (sugargoat) [inline ledger/sync/src/ledger_sync/ledger_sync_service_thread.rs:20]: Similarly here - should this be exposed?

**Review by sugargoat — APPROVED** [review]
LGTM! Might be worth exposing the magic numbers as params to ledger_sync

**Comment by eranrund** [comment]
> LGTM! Might be worth exposing the magic numbers as params to ledger_sync

I'll write a ticket to do it but I am not sure this is completely necessary. If we find ourselves tweaking those numbers then yes, but otherwise it would just add more passing of parameters around for no actual purpose.

**Pull request merged**